lib/ansible/modules/cloud/amazon/elasticsearch.py

```diff
@@ -0,0 +1,377 @@
+    returned: always
+    type: bool
+dedicated_master_type:
+    description: Instance type of dedicated master nodes.
+    returned: always
+    type: string
+    sample: "t2.medium.elasticsearch"
+dedicated_master_count:
+    description: Number of dedicated master nodes.
+    returned: always
+    type: int
+ebs_volume_type:
+    description: What EBS volume type nodes have.
+    returned: always
+    type: string
+    sample: "gp2"
+ebs_volume_size:
+    description: What size EBS volume each node has.
+    returned: always
+    type: int
+access_policies:
+    description: JSON string of the access policy of the cluster.
+    returned: always
+    type: string
+    sample: '{"Statement": [{"Action": "*", "Resource": "arn:aws:iam::0000000000000:user/you", "Effect": "Allow"}]}'
+"""
+
+import collections
+import json
+import time
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils import ec2
+
+try:
+    import botocore.exceptions
+except ImportError:
+    pass  # will be detected by ec2.HAS_BOTO3
+
+
+def _create_elasticsearch_domain(connection, module):
+    elasticsearch_cluster_config = {
+        'InstanceCount': module.params['instance_count'],
+        'InstanceType': module.params['instance_type'],
+        'ZoneAwarenessEnabled': module.params['zone_awareness'],
+        'DedicatedMasterEnabled': module.params['dedicated_master_enabled'],
+    }
+
+    if module.params['dedicated_master_enabled']:
+        elasticsearch_cluster_config['DedicatedMasterType'] = module.params['dedicated_master_type']
+        elasticsearch_cluster_config['DedicatedMasterCount'] = module.params['dedicated_master_count']
+
+    connection.create_elasticsearch_domain(
+        DomainName=module.params['domain'],
+        ElasticsearchVersion=module.params['version'],
+        ElasticsearchClusterConfig=elasticsearch_cluster_config,
+        EBSOptions={
+            'EBSEnabled': True,
+            'VolumeType': module.params['ebs_volume_type'],
+            'VolumeSize': module.params['ebs_volume_size'],
+        },
+        AccessPolicies=module.params['access_policies'],
+    )
+
+    return True
+
+
+def ensure_present(connection, module):
+    domain = module.params['domain']
+    elasticserach_domain = is_present(connection, domain)
+    if elasticserach_domain:
+        changed = False
+    else:
+        changed = True
+        _create_elasticsearch_domain(connection, module)
+        elasticserach_domain = is_present(connection, domain)
+    return changed, elasticserach_domain
+
+
+def is_present(connection, name):
+    """Get the current status of Elasticsearch domain"""
+    try:
+        response = connection.describe_elasticsearch_domain(
+            DomainName=name
+        )
+
+    except botocore.exceptions.ClientError as e:
+        if e.response['Error']['Code'] == 'ResourceNotFoundException':
+            return None
+        else:
+            raise
```

Comment:
Either this should have exception handling that calls fail_json or fail_json_aws (which would need module passed to this function) or everywhere that calls is_present should have the exception handling.